Issue #8: Added schema reference resolution

shahbaz-techprescient opened the pull request:
## Why

Added schema reference resolution when storing the schema in the generated Python classes.

## This PR

- does the above
- bumped the version to 0.0.5
- updated example schema and the example generated code

shahbaz-techprescient committed:
Issue #8: Added schema reference resolution

- Updated example schema and the generated code

drobert reviewed (approved):
I'd love to see a test verifying that this fix solves whatever problems existed before it, but the code itself looks fine (given I'm no python expert)

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,6 +408,25 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+        return
+
+    for field in schema.resolved_schema['fields']:
+        if isinstance(field['type'], str) and field['type'].find('.') >= 0:
```
When is `field['type']` not going to be a string? Also, isn't the latter part of this just `'.' in field['type']`?

shahbaz-techprescient replied:
In case of nested schema `field['type']` will be a dict.
Yes, I will change the latter part with your suggestion. It looks much more readable.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,6 +408,25 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+
+    for field in schema.resolved_schema['fields']:
+        if isinstance(field['type'], str) and field['type'].find('.') >= 0:
+            ref_schema_name = field['type'].split('.')[-1]
+            is_resolved = False
+            for s in schemas:
+                if s.name == ref_schema_name:
```
This is not going to work correctly for cases of shared names, right? i.e. `foo.message` and `bar.message` are going to get confused. here.

shahbaz-techprescient replied:
Good catch. Checking for {namespace}.{field_name}.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,6 +408,25 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+            is_resolved = False
+            for s in schemas:
+                if s.name == ref_schema_name:
+                    resolve_schema_reference(s, schemas)
```
This is going to mean that we're resolving things multiple times over, is there no caching mechanism we can employ?

shahbaz-techprescient replied:
Used a `set()` as a cache to store fully qualified name of resolved schemas.

hchasestevens replied:
I see - I think that's not quite the sort of caching I was imagining. I was thinking instead a dictionary where the keys are the fully qualified names and the values are the resolved JSON. 

Correct me if I'm wrong, but as-implemented, the caching mechanism/check on https://github.com/hchasestevens/avro2py/pull/9/files#diff-dd17a87ea1ee28b0a8ef80f33739b424462dfd309d0a7eb4d58b619f0ec386f3R414-R415 is not going to guarantee that schemas get resolved?

My understanding is that when we first encounter a reference to a new, unknown record (not in `RESOLVED_SCHEMAS`) that we will successfully invoke `resolve_schema_reference` and mutate, but that the second time we encounter it, in a new, unrelated record to our first encounter, because `RESOLVED_SCHEMAS` already has the fully-qualified name in it, `resolve_schema_reference` is going to just `return` without doing the necessary mutations; the reference will be left unresolved. Do you concur, and, if not, can you please write a test demonstrating that the above does not happen?

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,6 +408,25 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+                    field['type'] = s.resolved_schema
+                    is_resolved = True
+                    break
+            if not is_resolved:
+                raise ValueError(f'Unable to resolve schema reference {field["type"]} in {schema.namespace}.{schema.name}')
```
If we raise a `ValueError` here, we need to have some handling on the CLI level to let the user know what happened as opposed to the tool just failing.

shahbaz-techprescient replied:
The message in the exception will do that. The `cli` catch the exception and prints the stack trace. Also other parts of the code follow the same error handling.

Are you suggesting having a custom exception raised and handle it in cli ?

hchasestevens commented on avro2py/avro_types.py:
```diff
@@ -175,6 +176,7 @@ def parse_into_types(schema: Union[Dict[str, Any], str, List], parent_namespace:
     if type_name == 'record':
         return Record(
             original_schema=schema,
+            resolved_schema=schema,
```
This isn't going to work if these are the same schema, because when we mutate `resolved_schema` (which we do below), it's the same object that's going to be mutated. We either need to do a deepcopy on one of these, _not_ use mutation in `resolve_schema_reference` below, or get rid of `original_schema` entirely, if it's no longer being used.

shahbaz-techprescient replied:
I have used `deepcopy` for now.
My intention was to keep `original_schema` around if required for whatever reason later.
`original_schema` is not being used currently. We can delete it as well.

hchasestevens reviewed (changes requested):
Some comments inline, but we also need to have tests showing that the case that was previously failing is now succeeding

shahbaz-techprescient commented:
> I'd love to see a test verifying that this fix solves whatever problems existed before it, but the code itself looks fine (given I'm no python expert)

I have added one of the example of schema reference in the `example/schema.avsc` and generated the code in the `examples/marketplace` directory.
I will add a few tests.

shahbaz-techprescient committed:
Added cache in resolving the schema reference

-> Added tests
-> Accommodated other review comments

hchasestevens commented on avro2py/VERSION:
```diff
@@ -1 +1 @@
-0.0.4
\ No newline at end of file
+0.0.5
```
Because this constitutes a breaking change to the generated code, I would like to see this be 
```suggestion
0.1.0
```

hchasestevens commented on avro2py/avro_types.py:
```diff
@@ -175,6 +180,7 @@ def parse_into_types(schema: Union[Dict[str, Any], str, List], parent_namespace:
     if type_name == 'record':
         return Record(
             original_schema=schema,
+            resolved_schema=deepcopy(schema),
```
If `original_schema` is no longer being used, let's just rename that field to `schema` and not have two of these.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -168,7 +169,7 @@ def record_to_class(record: Record) -> ResolvedClassResult:
     class_body.append(
         ast.Assign(  # _original_schema = "..."
             targets=[ast.Name(id='_original_schema')],
```
Likewise, here, I would like to see this be
```suggestion
            targets=[ast.Name(id='_schema')],
```
as it is no longer the "original" from the AVSC, but our resolved/modified version.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,8 +409,29 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+
+    for field in schema.resolved_schema['fields']:
+        if isinstance(field['type'], str) and '.' in field['type']:
+            is_resolved = False
+            for s in schemas:
+                if field['type'] == s.fully_qualified_name():
+                    resolve_schema_reference(s, schemas)
+                    field['type'] = s.resolved_schema
+                    is_resolved = True
+                    break
+            if not is_resolved:
+                raise ValueError(f'Unable to resolve schema reference {field["type"]} in {schema.namespace}.{schema.name}')
```
Don't use this pattern with the sentinel value - use instead:
```suggestion
            for s in schemas:
                if field['type'] == s.fully_qualified_name():
                    resolve_schema_reference(s, schemas)
                    field['type'] = s.resolved_schema
                    break
            else:
                raise ValueError(f'Unable to resolve schema reference {field["type"]} in {schema.namespace}.{schema.name}')
```
See https://docs.python.org/3/reference/compound_stmts.html#for

shahbaz-techprescient replied:
I didn't know `else` can be used with `for`. Thanks for the suggestion.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -39,6 +39,7 @@
 
 
 NODE_CLASS_CONVERTERS = {}
+RESOLVED_SCHEMAS = set()  # To cache resolved schemas
```
I would like to see this not be a module-level global and instead just be maintained within the appropriate function, and passed through as an argument to `resolve_schema_reference` below.

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,8 +409,29 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+
 def populate_namespaces(schemas: List[Record]) -> Generator[Tuple[str, ast.Module], None, None]:
     """Convert internal Record representations into renderable AST module nodes."""
+    RESOLVED_SCHEMAS = set()  # initialize (or invalidate existing) the cache.
```
I don't think this will work like you expect it to - this looks to be shadowing the module-level variable? Consider the following code:

```python
In [1]: RESOLVED_SCHEMAS = set()

In [2]: def foo():
   ...:     RESOLVED_SCHEMAS = set()  # "reset" module-level cache
   ...:

In [3]: def bar():
   ...:     RESOLVED_SCHEMAS.add(1)
   ...:

In [4]: RESOLVED_SCHEMAS
Out[4]: set()

In [5]: bar()

In [6]: RESOLVED_SCHEMAS
Out[6]: {1}

In [7]: foo()

In [8]: RESOLVED_SCHEMAS  # was not reset
Out[8]: {1}
```

To fix this, you would have to have `global RESOLVED_SCHEMAS` come before this in this function, but, per my comment above, let's make this internal to this function exclusively (do not uppercase) and pass it in explicitly to `resolve_schema_reference` in https://github.com/hchasestevens/avro2py/pull/9/files#diff-dd17a87ea1ee28b0a8ef80f33739b424462dfd309d0a7eb4d58b619f0ec386f3R456-R457 below, so we don't have to worry about global state whatsoever.

hchasestevens commented on tests/test_codegen.py:
```diff
@@ -38,10 +38,31 @@ def test_bridge_namespaces(from_namespace, to_namespace, name, expected):
+    parsed_schemas = [parse_into_types(schema) for schema in TEST_SCHEMAS]
+    list(codegen.populate_namespaces(parsed_schemas))  # since it produces generator
+
+    ad_connected_schema = None
+    for ps in parsed_schemas:
+        if ps.fully_qualified_name() == 'marketprice.messages.events.advertising.AdConnected':
+            ad_connected_schema = ps
+            break
+    assert ad_connected_schema is not None
```
I would again use
```suggestion
    for ps in parsed_schemas:
        if ps.fully_qualified_name() == 'marketprice.messages.events.advertising.AdConnected':
            ad_connected_schema = ps
            break
    else:
        raise Exception("couldn't find AdConnected schema")
```
rather than the sentinel, but in this particular circumstance, please instead just do
```suggestion
    ad_connected_schema, = (
        schema
        for schema in parsed_schemas
        if schema.fully_qualified_name() == 'marketprice.messages.events.advertising.AdConnected'
    )
```
, which will have the same effect _and_ ensure that the schema found is unique.

shahbaz-techprescient replied:
I like this pattern. I will do the change. I am glad I chose to fix `avro2py` 🙂 .

hchasestevens commented on tests/test_codegen.py:
```diff
@@ -38,10 +38,31 @@ def test_bridge_namespaces(from_namespace, to_namespace, name, expected):
         assert isinstance(ast.parse(source), ast.Module)
+
+
+def test_schema_reference_resolution():
+    parsed_schemas = [parse_into_types(schema) for schema in TEST_SCHEMAS]
```
This test needs a docstring explaining what properties it's actually trying to ensure hold.

hchasestevens commented on tests/test_codegen.py:
```diff
@@ -38,10 +38,31 @@ def test_bridge_namespaces(from_namespace, to_namespace, name, expected):
+            break
+    assert ad_connected_schema is not None
+
+    for field in ad_connected_schema.resolved_schema['fields']:
+        if field['name'] == 'metadata':
+            assert isinstance(field['type'], dict)
+            assert field['type']['name'] == 'EventMetadata'
+            assert field['type']['namespace'] == 'marketprice.messages.events'
+        elif field['name'] == 'foo_metadata':
+            assert isinstance(field['type'], dict)
+            assert field['type']['name'] == 'EventMetadata'
+            assert field['type']['namespace'] == 'foo.events'
```
Per https://docs.pytest.org/en/reorganize-docs/new-docs/user/assert_statements.html , tests should only have one `assert`. Can we perhaps use a `pytest.mark.parametrize` decorator to make this into two tests, one for `'metadata'` and one for `'foo_metadata'`, with the corresponding expected namespace values (`'marketprice.messages.events'` and `'foo.events'`)? This can then just be:
```suggestion
    assert all(
        field['type']['namespace'] == expected_namespace
        for field in ad_connected_schema.resolved_schema['fields']
        if field['name'] == expected_name
    )
```

shahbaz-techprescient replied:
I have not used `pytest` much but I like the `parametrize` functionality.
I have applied you suggestion in a different way. Hope this is way is also ok

hchasestevens commented on avro2py/codegen.py:
```diff
@@ -408,8 +409,29 @@ def visit_AnnAssign(self, node: ast.AnnAssign) -> ast.AnnAssign:
+        return
+
+    for field in schema.resolved_schema['fields']:
+        if isinstance(field['type'], str) and '.' in field['type']:
```
If we're only looking at types that are strings, doesn't that mean we're going to fail to resolve references in nested/complex types (unions, arrays, maps, etc.), e.g. https://github.com/hchasestevens/avro2py/blob/master/example/schema.avsc#L1407-L1415 or https://github.com/hchasestevens/avro2py/blob/master/example/schema.avsc#L697-L700 ?

hchasestevens replied:
I think this problem is still outstanding and needs to be addressed.

hchasestevens commented on avro2py/avro_types.py:
```diff
@@ -60,12 +61,16 @@ class LogicalType(NamedTuple):
     doc: Optional[str] = None
     aliases: List[str] = []
 
+    def fully_qualified_name(self):
+        """Return fully qualified name."""
+        return f'{self.namespace}.{self.name}'
```
Rather than a method, let's just make this an `@property`:
```suggestion
    @property
    def fully_qualified_name(self):
        """Return fully qualified name."""
        return f'{self.namespace}.{self.name}'
```
Additionally, if we're going to introduce this, we should be using it in all loci that require the fully-qualified name; https://github.com/hchasestevens/avro2py/blob/master/avro2py/codegen.py#L312 , for example.

shahbaz-techprescient replied:
Sure. But since line #312 is applicable to both `Record` and `Enum`. I have added this property in `Enum` as well.

shahbaz-techprescient committed:
Bump version

Co-authored-by: H. Chase Stevens <chase@chasestevens.com>

shahbaz-techprescient committed:
Replace original_schema and resolved_schema with just one var schema.

-> Modified tests

shahbaz-techprescient committed:
Removed the global cache variable.

Added test to validate caching does not leave second occurence of types unresolved

shahbaz-techprescient committed:
Re-generated classes in the example folder

drobert commented:
Open question here (which may be 'bad news bears'), but per an ongoing conversation in slack:

Older versions of Confluent Schema Registry (and only < 5.1 are open licensing) do not support 'schema references'. We currently use schema references so some additional testing is underway, but I want to highlight an example and some notes on how java codegen operates.

Consider this source schema:
```json
[
  {
    "type": "record",
    "namespace": "example.avro",
    "name": "Money",
    "doc": "An amount of a particular currency",
    "fields": [
      {
        "name": "amount",
        "type": {
          "type": "bytes",
          "logicalType": "decimal",
          "scale": 2,
          "precision": 19
        }
      },
      {
        "name": "currency",
        "type": "string",
        "doc": "3-character ISO 4217 currency code"
      }
    ]
  },
  {
    "type": "record",
    "namespace": "example.avro",
    "name": "Twofer",
    "doc": "Using a schema reference twice",
    "fields": [
      {
        "name": "event_id",
        "type": {
          "type": "string",
          "logicalType": "uuid"
        }
      },
      {
        "name": "in",
        "type": "example.avro.Money"
      },
      {
        "name": "out",
        "type": "example.avro.Money"
      }
    ]
  }
]
```

After java codgen, the generated schema (embedded in the resulting `Twofer` type) looks like this:
```json
{
  "type": "record",
  "name": "Twofer",
  "namespace": "example.avro",
  "doc": "Using a schema reference twice",
  "fields": [
    {
      "name": "event_id",
      "type": {
        "type": "string",
        "logicalType": "uuid"
      }
    },
    {
      "name": "in",
      "type": {
        "type": "record",
        "name": "Money",
        "doc": "An amount of a particular currency",
        "fields": [
          {
            "name": "amount",
            "type": {
              "type": "bytes",
              "logicalType": "decimal",
              "scale": 2,
              "precision": 19
            }
          },
          {
            "name": "currency",
            "type": {
              "type": "string",
              "avro.java.string": "String"
            },
            "doc": "3-character ISO 4217 currency code"
          }
        ]
      }
    },
    {
      "name": "out",
      "type": "Money"
    }
  ]
}
```

Noticing in this example, the first use of `Money` is in-lined while subsequent ones are by reference. Unclear to me if multiple in-lines of the same type would function, and we're still testing what effect any of this has on the schema registry.

**Update**: confirmed that the java-compiler-rewritten version of this schema is compatible with our schema registry version. I think ideally our python-codgen schema format would take the same approach

drobert reviewed (changes requested):
Rescinding my 'approve' in light of concerns about references and inlining. I'd like to see an example of this schema being codegen'd and see that the resulting schema remains valid (that is, does not inline the `Money` field twice):

```json
// source schema
[
  {
    "type": "record",
    "namespace": "common.avro",
    "name": "Money",
    "doc": "An amount of a particular currency",
    "fields": [
      {
        "name": "amount",
        "type": {
          "type": "bytes",
          "logicalType": "decimal",
          "scale": 2,
          "precision": 19
        }
      },
      {
        "name": "currency",
        "type": "string",
        "doc": "3-character ISO 4217 currency code"
      }
    ]
  },
{
    "type": "record",
    "namespace": "example.avro",
    "name": "Twofer",
    "doc": "Using a schema reference twice",
    "fields": [
      {
        "name": "event_id",
        "type": {
          "type": "string",
          "logicalType": "uuid"
        }
      },
      {
        "name": "in",
        "type": "common.avro.Money"
      },
      {
        "name": "out",
        "type": "common.avro.Money"
      }
    ]
  }
]

// expected resulting (codegen'd) schema
{
  "type": "record",
  "name": "Twofer",
  "namespace": "example.avro",
  "doc": "Using a schema reference twice",
  "fields": [
    {
      "name": "event_id",
      "type": {
        "type": "string",
        "logicalType": "uuid"
      }
    },
    {
      "name": "in",
      "type": {
        "type": "record",
        "name": "Money",
        "namespace": "common.avro",
        "doc": "An amount of a particular currency",
        "fields": [
          {
            "name": "amount",
            "type": {
              "type": "bytes",
              "logicalType": "decimal",
              "scale": 2,
              "precision": 19
            }
          },
          {
            "name": "currency",
            "type": {
              "type": "string",
              "avro.java.string": "String"
            },
            "doc": "3-character ISO 4217 currency code"
          }
        ]
      }
    },
    {
      "name": "out",
      "type": "common.avro.Money"
    }
  ]
}
```

shahbaz-techprescient commented:
No longer required, as there is a workaround implemented.
